### PR TITLE
Add `expandRecurringEvent` API for Recurring Event Expansion

### LINKED ISSUE #113: Repeating Events hard to handle
I wish for a better way to handle repeaing events in the parser like parsing the repeating events as a seperate object and not having only one with the attributes...idk but it's hard to detect when a repeated event takes place. You only have the original wich says yea...i repeat but you have to manage it yourself..

### PR DESCRIPTION
**Pull request by KristjanESPERANTO**
This PR implements a new public API `expandRecurringEvent()` that expands recurring calendar events into discrete instances within a specified date range. This eliminates the need for consumers to implement complex recurring event logic themselves.

Fixes #113

## Motivation

Currently, consumers (like **ioBroker.ical** and **MagicMirror**) need to implement their own complex logic to:
- Expand RRULE-based recurring events into individual occurrences
- Handle EXDATE exclusions (cancelled instances)
- Apply RECURRENCE-ID overrides (modified instances)
- Deal with timezone and DST issues for all-day events
- Filter events within date ranges

This leads to:
- **Duplicated code** across multiple consumers
- **Inconsistent behavior** between different implementations
- **Complex maintenance burden** when iCalendar spec edge cases arise
- **Bugs in DST handling** and timezone conversions

## Solution

This PR adds a single, well-tested function that handles all recurring event expansion logic:

```javascript
const instances = ical.expandRecurringEvent(event, {
  from: new Date('2025-01-01'),
  to: new Date('2025-01-31'),
  includeOverrides: true,
  excludeExdates: true,
  expandOngoing: false
});

// Returns:
// [
//   {
//     start: Date,
//     end: Date,
//     summary: string,
//     isFullDay: boolean,
//     isRecurring: boolean,
//     isOverride: boolean,
//     event: VEvent
//   },
//   ...
// ]
```

## Benefits

- Removes the need for custom recurring-event expansion logic in consumers.
- Centralizes EXDATE/RECURRENCE-ID handling and DST-safe full-day behavior.
- Provides a single, consistent API for recurring and non-recurring events.

### Function Signature

```typescript
function expandRecurringEvent(
  event: VEvent,
  options: {
    from: Date;                  // Start of date range (inclusive)
    to: Date;                    // End of date range (inclusive)
    includeOverrides?: boolean;  // Apply RECURRENCE-ID overrides (default: true)
    excludeExdates?: boolean;    // Filter out EXDATE exclusions (default: true)
    expandOngoing?: boolean;     // Include events that started before 'from' but are still ongoing (default: false)
  }
): EventInstance[];

interface EventInstance {
  start: Date;          // Instance start time
  end: Date;            // Instance end time
  summary: string;      // Event title (from override or original event)
  isFullDay: boolean;   // Whether this is an all-day event
  isRecurring: boolean; // Whether this came from a recurring event
  isOverride: boolean;  // Whether this instance has a RECURRENCE-ID override
  event: VEvent;        // The event object (override or original)
}
```

### Key Features

- **Works for both recurring AND non-recurring events** - unified API
- **RRULE expansion** using the parsed `event.rrule` (as provided by node-ical)
- **EXDATE filtering** - excluded instances are automatically removed
- **RECURRENCE-ID overrides** - modified instances use override data
- **DST-safe** - all-day events use UTC component extraction to avoid timezone shifts
- **Date range filtering** - only returns instances within specified range
- **Chronologically sorted** - instances are returned in time order
- **Ongoing event support** - optionally include events that started earlier

#### Basic Usage
```javascript
const ical = require('node-ical');

const events = await ical.fromURL('https://example.com/calendar.ics');
const dailyStandup = events['daily-standup@example.com'];

const nextMonth = ical.expandRecurringEvent(dailyStandup, {
  from: new Date('2025-02-01'),
  to: new Date('2025-02-28')
});

console.log(`Found ${nextMonth.length} standup meetings in February`);
```

#### With EXDATE Exclusions
```javascript
// Event has EXDATE for 2025-02-14 (Valentine's Day - no meeting)
const instances = ical.expandRecurringEvent(event, {
  from: new Date('2025-02-01'),
  to: new Date('2025-02-28'),
  excludeExdates: true  // Skip excluded dates
});

// Feb 14 instance will be automatically excluded
```

#### With RECURRENCE-ID Overrides
```javascript
// Event has RECURRENCE-ID override on 2025-02-21 (different time/summary)
const instances = ical.expandRecurringEvent(event, {
  from: new Date('2025-02-01'),
  to: new Date('2025-02-28'),
  includeOverrides: true  // Apply modified instances
});

// Feb 21 instance will use override summary and time
instances.forEach(instance => {
  console.log(instance.isOverride ? '[MODIFIED]' : '[NORMAL]', instance.summary);
});
```

#### Ongoing Events
```javascript
// Include multi-day events that started before range
const instances = ical.expandRecurringEvent(vacation, {
  from: new Date('2025-02-15'),
  to: new Date('2025-02-20'),
  expandOngoing: true  // Include events that started earlier but are still active
});

// Will include a vacation that started on Feb 10 and ends Feb 18
```

### Architecture
- **6 helper functions** to keep complexity low (max complexity: 10 per function)
- **Separation of concerns**: date key generation, DST handling, event processing, range checking
- **Type-safe**: Full TypeScript definitions included

### Helper Functions
1. `generateDateKey(date, isFullDay)` - Creates lookup keys for EXDATE/RECURRENCE-ID matching
2. `createLocalDateFromUTC(utcDate)` - Extracts UTC components for DST-safe all-day events
3. `calculateEndTime(start, eventData, isFullDay, isRecurring)` - Computes end time with proper fallbacks
4. `processNonRecurringEvent(event, options)` - Handles single events
5. `processRecurringInstance(date, event, options)` - Processes each recurrence
6. `isInstanceInRange(instance, from, to, expandOngoing)` - Range checking for date filtering

### DST Handling
Full-day events are processed using UTC component extraction to avoid DST shifts:
```javascript
// For full-day events, extract UTC components
const year = utcDate.getUTCFullYear();
const month = utcDate.getUTCMonth();
const day = utcDate.getUTCDate();
const localDate = new Date(year, month, day);
```

This ensures that "2025-03-10" remains "2025-03-10" regardless of DST transitions.

### Limitations / Notes
- **`expandOngoing` look-back window**: For recurring events, the implementation looks back by the event duration (derived from DTSTART/DTEND or default rules) to capture ongoing instances.
- **Full-day range comparison**: Full-day events are compared by calendar date (local date), not by timestamp, to avoid DST-related shifts.

### Test Coverage
- **111 tests passing** (24 new + 87 existing)
- **Edge cases tested**: DST transitions, timezones, EXDATE, RECURRENCE-ID, duration calculations, input validation

### New Test Suite
```
expandRecurringEvent
  Basic functionality
    ✓ should expand a daily recurring event
    ✓ should handle non-recurring events
    ✓ should respect date range boundaries
  Return structure
    ✓ should return correct instance structure
    ✓ should set isRecurring correctly
    ✓ should handle all-day events
    ✓ should calculate end times
  EXDATE handling
    ✓ should exclude EXDATE instances
    ✓ should include EXDATE when excludeExdates=false
  RECURRENCE-ID handling
    ✓ should apply RECURRENCE-ID overrides
    ✓ should mark overrides with isOverride flag
    ✓ should skip overrides when includeOverrides=false
  DST handling
    ✓ should handle all-day events across DST transitions
    ✓ should handle timed events across DST transitions
  Duration handling
    ✓ should handle events with explicit end times
    ✓ should handle events with duration
    ✓ should default to zero duration for timed events
  Edge cases
    ✓ should handle events with no summary
    ✓ should handle expandOngoing option for non-recurring events
    ✓ should handle expandOngoing option for recurring events
    ✓ should return empty array for out-of-range events
  Input validation
    ✓ should throw TypeError for invalid from date
    ✓ should throw TypeError for invalid to date
    ✓ should throw RangeError when from is after to
```

## Breaking Changes

**None.** This PR only adds new functionality without modifying existing APIs.

## Summary by CodeRabbit

* **New Features**
  * Added an event-expansion utility to produce individual event instances from recurring and non-recurring VEVENTs within a date range. DST-safe, supports overrides, EXDATE handling, ongoing-event inclusion, and returns sorted instances.

* **Documentation**
  * New "Expanding recurring events" section with usage examples and option descriptions (section duplicated).

* **Tests**
  * Comprehensive test suite covering expansion behavior, DST, overrides, exclusions, durations, edge cases.

* **Examples**
  * Example scripts updated to use the new expansion API for simpler instance iteration.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: test/expand-recurring-event.test.js [hunk 1]
```diff
@@ -0,0 +1,473 @@
...
+describe('expandRecurringEvent', () => {
+  describe('Basic functionality', () => {
+    it('should expand a simple daily recurring event', () => {
+      const events = ical.sync.parseFile(path.join(__dirname, 'test_daily_recurring.ics'));
+      const event = Object.values(events).find(event => event.type === 'VEVENT');
+
+      const instances = ical.expandRecurringEvent(event, {
+        from: new Date('2025-01-01'),
+        to: new Date('2025-01-07'),
+      });
```
[the remark this answers is not in the document]
Comment (KristjanESPERANTO) [reply test/expand-recurring-event.test.js:15]: Done with 79c153e3bcc36ffd956b48157e1dbb91ce2287d6.

File: node-ical.js [hunk 2]
```diff
@@ -237,6 +237,245 @@ autodetect.parseICS = function (data, cb) {
...
+  let start = date;
+
+  // For full-day events, extract UTC components to avoid DST issues
+  if (isFullDay) {
+    start = createLocalDateFromUTC(date);
+  }
+
+  // For recurring events, use override duration when available; otherwise use base duration
+  const end = calculateEndTime(start, instanceEvent, isFullDay, baseDurationMs);
+
```
[the remark this answers is not in the document]
Comment (KristjanESPERANTO) [reply node-ical.js:395]: Good catch! Fixed and added test for it. See 79c153e3bcc36ffd956b48157e1dbb91ce2287d6.

File: node-ical.js [hunk 3]
```diff
@@ -237,6 +237,245 @@ autodetect.parseICS = function (data, cb) {
...
+ */
+function processNonRecurringEvent(event, options) {
+  const {from, to, expandOngoing} = options;
+  const isFullDay = event.datetype === 'date' || Boolean(event.start?.dateOnly);
+  const baseDurationMs = getEventDurationMs(event, isFullDay);
+
+  // Ensure we have a proper Date object
+  const eventStart = event.start instanceof Date ? event.start : new Date(event.start);
+  const eventEnd = calculateEndTime(eventStart, event, isFullDay, baseDurationMs);
+
```
[the remark this answers is not in the document]
Comment (KristjanESPERANTO) [reply node-ical.js:330]: Good catch! Fixed and added test for it. See 79c153e3bcc36ffd956b48157e1dbb91ce2287d6.

**Comment by KristjanESPERANTO** [comment]
@coderabbitai review

File: node-ical.js [hunk 4]
```diff
@@ -237,6 +237,264 @@ autodetect.parseICS = function (data, cb) {
...
+function generateDateKey(date, isFullDay) {
+  if (isFullDay) {
+    const year = date.getUTCFullYear();
+    const month = date.getUTCMonth() + 1;
+    const day = date.getUTCDate();
+    return `${year}-${String(month).padStart(2, '0')}-${String(day).padStart(2, '0')}`;
+  }
+
+  return date.toISOString();
+}
```
[the remark this answers is not in the document]
Comment (KristjanESPERANTO) [reply node-ical.js:258]: Done with 5c532e47127b6b8033ba046b87d7693149e06a99 :slightly_smiling_face:

**Comment by KristjanESPERANTO** [comment]
@coderabbitai review

**Comment by KristjanESPERANTO** [comment]
I think coderabbit is satisfied now :slightly_smiling_face:

@jens-maus, what do you think of this new feature?

**Comment by jens-maus** [comment]
> @jens-maus, what do you think of this new feature?

This is a great feature! Looking forward using it in the various side projects! Thanks for implementing it.

**Comment by KristjanESPERANTO** [comment]
@jens-maus I'm glad you like the feature :slightly_smiling_face: Please make a new release when you get a chance.

**Comment by KristjanESPERANTO** [comment]
I see  the Windows builds for Node 18/20 have failed due to a GitHub Actions infrastructure issue (they failed instantly, before running any code) :thinking:  Re-running after a couple of hours may fix it. But this is not an issue of our code.

**Comment by jens-maus** [comment]
> I see the Windows builds for Node 18/20 have failed due to a GitHub Actions infrastructure issue (they failed instantly, before running any code) 🤔 Re-running after a couple of hours may fix it. But this is not an issue of our code.

Yes, I also just noticed that. Let's wait until tomorrow until GitHub has fixed the issue, then I will prepare a new release ASAP.

**Comment by KristjanESPERANTO** [comment]
Many thanks! :-)
